File: ide/src/main/java/org/dau/util/Encoders.java
package org.dau.util;

import java.nio.ByteBuffer;
import java.util.Base64;

public final class Encoders {

  private Encoders() {}

  public static final Base64.Encoder ID_ENCODER = Base64.getUrlEncoder().withoutPadding();

  public static String generateId(Object o) {
    int timestamp = (int) (System.currentTimeMillis() / 1000L);
    int hash = System.identityHashCode(o);
    var bytes = ByteBuffer.allocate(8).putInt(0, timestamp).putInt(4, hash).array();
    return ID_ENCODER.encodeToString(bytes);
  }
}

File: ide/src/main/kotlin/org/dau/ui/utils/Encoders.kt
package org.dau.ui.utils

import java.nio.ByteBuffer
import java.time.LocalDate
import java.util.*

object Encoders {

  val ID_ENCODER = Base64.getUrlEncoder().withoutPadding()

  @JvmStatic
  fun generateId(o: Any): String {
    val ldt = LocalDate.now()
    val timePart = String.format("%02d%02d%02d", ldt.year % 100, ldt.monthValue, ldt.dayOfMonth)
    val hash = System.identityHashCode(o)
    val bytes = ByteBuffer.allocate(4).putInt(0, hash).array()
    return timePart + "_" + ID_ENCODER.encodeToString(bytes)
  }
}